23.0.0.12-beta staging (#3457)
OpenLiberty/blogs: posts/2023-11-28-23.0.0.12-beta.adoc

# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/lauracowen
author_github: https://github.com/lauracowen
seo-title: Preview of Jakarta Data Milestone 1 in Open Liberty 23.0.0.12-beta - OpenLiberty.io
seo-description: Annotate lifecycle methods in a standardized way with Jakarta Data. Try Milestone 1 now and give feedback on the spec so far.
blog_description: Annotate lifecycle methods in a standardized way with Jakarta Data. Try Milestone 1 now and give feedback on the spec so far.
open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
---
= Preview of Jakarta Data Milestone 1 in Open Liberty 23.0.0.12-beta
Laura Cowen <https://github.com/lauracowen>
:imagesdir: /
//Blank line here is necessary before starting the body of the post.

Open Liberty 23.0.0.12-beta contains a preview of the new Jakarta Data specification for Jakarta EE, as it currently stands at Milestone 1. Milestone 1 provides the capability to annotate lifecycle methods such as `Insert`, `Delete`, and more. You can try it out and give feedback on the specification so far.

> In this beta we have also added the ability to configure the duration of the _quiesce_ stage when the Liberty runtime shutsdown. This means that, for services that need longer than the default 30 seconds to complete the processing of requests, you can set the quiesce stage to a longer duration.

Reviewer: ```suggestion
Also in this beta, you can configure the _quiesce_ stage when the Liberty runtime shuts down to be longer than the default 30 seconds. This update is useful for services that need more time to finish processing requests.
```

The link:/[Open Liberty] 23.0.0.12-beta includes the following beta features (along with link:/docs/latest/reference/feature/feature-overview.html[all GA features]):

* <<jdata, Preview of Jakarta Data (Milestone 1)>>
* <<timeout, Configurable Quiesce Timeout>>

See also link:/blog/?search=beta&key=tag[previous Open Liberty beta blog posts].

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/26759
// Contact/Reviewer: njr-11
// // // // // // // // 
[#jdata]
== Preview of Jakarta Data (Milestone 1)

> Jakarta Data is a new link:https://jakarta.ee/specifications/data/1.0/[Jakarta EE specification] being developed in the open that aims to standardize the popular data link:https://deviq.com/design-patterns/repository-pattern[repository pattern] across a variety of providers. Open Liberty includes the Jakarta Data 1.0 Milestone 1 release, which adds the ability to annotatively compose custom lifecycle methods, covering `Insert`, `Update`, `Save`, and `Delete` operations. The Open Liberty beta includes a test implementation of Jakarta Data that we are using to experiment with proposed specification features so that developers can try out these features and provide feedback to influence the Jakarta Data 1.0 specification as it continues to be developed after Milestone 1. The test implementation currently works with relational databases and operates by redirecting repository operations to the built-in Jakarta Persistence provider.

Reviewer 2: I don't think we should link to another vendor's documentation here. After the initial paragraph it gets into specifics of that vendor's own solution.
```suggestion
Jakarta Data is a new link:https://jakarta.ee/specifications/data/1.0/[Jakarta EE specification] being developed in the open that aims to standardize the popular data repository pattern across a variety of providers. Open Liberty includes the Jakarta Data 1.0 Milestone 1 release, which adds the ability to annotatively compose custom lifecycle methods, covering `Insert`, `Update`, `Save`, and `Delete` operations. The Open Liberty beta includes a test implementation of Jakarta Data that we are using to experiment with proposed specification features so that developers can try out these features and provide feedback to influence the Jakarta Data 1.0 specification as it continues to be developed after Milestone 1. The test implementation currently works with relational databases and operates by redirecting repository operations to the built-in Jakarta Persistence provider.
```

Start by defining an entity class that corresponds to your data. With relational databases, the entity class corresponds to a database table and the entity properties (public methods and fields of the entity class) generally correspond to the columns of the table. An entity class can be:

* annotated with `jakarta.persistence.Entity` and related annotations from Jakarta Persistence
* a Java class without entity annotations, in which case the primary key is inferred from an entity property named `id` or ending with `Id` and an entity property named `version` designates an automatically incremented version column.

You define one or more repository interfaces for an entity, annotate those interfaces as `@Repository` and inject them into components via `@Inject`. The Jakarta Data provider supplies the implementation of the repository interface for you.

The following example shows a simple entity:

[source,java]
----
@Entity
public class Product {
    @Id
    public long id;

    public String name;

    public float price;

    @Version
    public long version;
}
----

> The following example shows a repository that defines operations relating to the entity. Your repository interface can inherit from built-in interfaces such as `BasicRepository` and `CrudRepository` to gain a variety of general purpose repository methods for inserting, updating, deleting, and querying for entities, but in this case we will define all of the methods ourselves, taking advantage of the new lifecycle annotations:

Reviewer: ```suggestion
The following example shows a repository that defines operations that relate to the entity. Your repository interface can inherit from built-in interfaces, such as `BasicRepository` and `CrudRepository`, to gain a variety of general purpose repository methods for inserting, updating, deleting, and querying for entities. However, in this case, we will define all of the methods ourselves by using the new lifecycle annotations:
```

    @Inject
    Products products;

    protected void doGet(HttpServletRequest req, HttpServletResponse resp)
            throws ServletException, IOException {
        // Insert:
        Product prod = ...
        prod = products.add(prod);

        // Update:
        prod.price = prod.price + 1.00;
        if (products.update(prod))
            System.out.println("Successfully increased the price.");
        else {
            // someone else either removed the product or updated its version before we could
            prod = products.find(productId).orElseThrow();
            ...
        }

        // Request only the first 20 results on a page, ordered by price, then name, then id:
        Pageable pageRequest = Pageable.size(20).sortBy(Sort.desc("price"), Sort.asc("name"), Sort.asc("id"));
        Page<Product> page1 = products.findByNameIgnoreCaseContains(searchFor, pageRequest);
        ...
    }
}
----

    
// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC> 

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/26724
// Contact/Reviewer: jimblye
// // // // // // // // 
[#timeout]

> == Configurable quiesce timeout for the Liberty runtime
> 
> Liberty has a _quiesce_ stage when shutting down the Liberty runtime, which prevents services from accepting new requests and allows time for services to process existing requests. The quiesce stage has always been a fixed 30 seconds time period. This quiesce time period is now configurable.
> 
> Previously, in some cases, the 30-seconds quiesce period was not long enough time for services to finish processing existing requests. So you can now increase the quiesce timeout if necessary.
> 
> To configure the quiesce timeout, add the new `quiesceTimeout` attribute to the `applicationManager` element in the `server.xml` file:
> 
> [source,xml]
> ----
> <applicationManager quiesceTimeout=“1m30s”/>
> ----
> 
> The timeout value is a positive integer followed by a unit of time, which can be hours (`h`), minutes (`m`), or seconds (`s`). For example, specify 30 seconds as `30s`. You can include multiple units in a single entry. For example, `1m30s` is equivalent to 90 seconds. The minimum `quiesceTimeout` value is 30 seconds. If you specify a shorter length of time, the value `30s` is used.
> 
> For more info, see the link:https://www.ibm.com/docs/en/was-liberty/base?topic=configuration-applicationmanager[Application Manager (applicationManager)] element docs.

Reviewer 3: ```suggestion
== Configurable quiesce timeout for the Liberty runtime

Liberty has a _quiesce_ stage when shutting down the Liberty runtime, which prevents services from accepting new requests and allows time for services to process existing requests. The quiesce stage has always been a fixed 30 seconds time period. This quiesce time period is now configurable.

Previously, in some cases, the 30-seconds quiesce period was not long enough time for services to finish processing existing requests. So you can now increase the quiesce timeout if necessary.

To configure the quiesce timeout, add the new `quiesceTimeout` attribute to the `executor` element in the `server.xml` file:

[source,xml]
----
<executor quiesceTimeout=“1m30s”/>
----

The timeout value is a positive integer followed by a unit of time, which can be hours (`h`), minutes (`m`), or seconds (`s`). For example, specify 30 seconds as `30s`. You can include multiple units in a single entry. For example, `1m30s` is equivalent to 90 seconds. The minimum `quiesceTimeout` value is 30 seconds. If you specify a shorter length of time, the value `30s` is used.
```

Reviewer 3: @jimblye does the above look correct?

Reviewer 4: looks good

Reviewer: not sure whether we should link to the doc here for the beta since it wont include any info about `quiesceTimeout` until the feature is GA
